### PR TITLE
gss_data refcounting fixes

### PR DESCRIPTION
**Pull request by dang**

### CODE REVIEW

File: ntirpc/rpc/gss_internal.h [hunk 1]
```diff
@@ -119,6 +119,9 @@ unref_svc_rpc_gss_data(struct svc_rpc_gss_data *gd)
 	/* if refcnt is 0, gd is not reachable */
 	if (unlikely(atomic_dec_uint32_t(&gd->refcnt) == 0)) {
 		svcauth_gss_destroy(gd->auth);
+		/* gd was freed.  We can't unlock the lock.  But, it's safe to
+		 * destroy a lock if you own it.  */
+		return;
```
Comment (malahal) [inline ntirpc/rpc/gss_internal.h:123]: Not sure if I understand the comment "But, it's safe to destroy a lock if you own it.".  In fact, svcauth_gss_destroy() calls unlock and destroy over gd_lock. This is bad idiom, why we need lock around svcauth_gss_destroy() ? Can we completely remove as the atomic refcount should ensure that there is only one thread calling svcauth_gss_destroy() ?

File: src/svc_auth_gss.c [hunk 2]
```diff
@@ -660,13 +654,11 @@ _svcauth_gss(struct svc_req *req, bool *no_dispatch)
 	}
 gd_free:
 	mutex_unlock(&gd->lock);
-	if (gd_hashed) {
-		unref_svc_rpc_gss_data(gd);
-		gd_hashed = false;
-	}
+	unref_svc_rpc_gss_data(gd);
```
Comment (malahal) [inline src/svc_auth_gss.c]: We should hold a ref in a success case. The macro conversion to gotos lost that. We should only call unref in the case of a failure.

File: ntirpc/rpc/gss_internal.h (hunk 1 above)
Comment (malahal) [reply ntirpc/rpc/gss_internal.h:123]: We call svcauth_gss_destroy() directly from unref_svc_rpc_gss_data() with refcounts but we also pass svcauth_gss_destroy() as a function pointer in svc_auth_gss_ops structure. Who calls it and when? Can svcauth_gss_destroy() get called with active refcount on gd?

Comment (mattbenjamin) [reply ntirpc/rpc/gss_internal.h:123]: yes, now that there are assured to be no refs--and no more can be acquired--it's safe to destroy it

**Review by mattbenjamin — APPROVED** [review]
lgtm

**Pull request merged**